File: web/app/src/lib/mcpConfig.ts
// Pure helpers for the hosted MCP surface (Streamable HTTP). Shared by the
// Settings developer section and its tests so the displayed server URL and the
// copied Claude config can never drift apart.

/** Canonical hosted MCP endpoint path appended to the API base. */
export const MCP_ENDPOINT_PATH = '/v1/mcp';

/**
 * Canonical hosted MCP URL derived from `apiBase` (e.g.
 * `NEXT_PUBLIC_API_BASE_URL`), with trailing slashes normalized away.
 */
export function hostedMcpUrl(apiBase: string): string {
  return `${apiBase.replace(/\/+$/, '')}${MCP_ENDPOINT_PATH}`;
}

/**
 * The Claude Code `~/.claude.json` `mcpServers` config JSON for the hosted
 * Streamable HTTP endpoint — `type: "http"` plus a Bearer header, no local
 * transport. The same string is rendered and copied. (Claude Desktop adds
 * remote servers via Settings → Connectors, not a JSON file.)
 */
export function hostedMcpConfigJson(mcpUrl: string): string {
  return `{
  "mcpServers": {
    "omi": {
      "type": "http",
      "url": ${JSON.stringify(mcpUrl)},
      "headers": {
        "Authorization": "Bearer <key>"
      }
    }
  }
}`;
}

File: app/lib/utils/mcp_config.dart
import 'dart:convert';

// Pure helpers for the hosted MCP surface (Streamable HTTP).
// Flutter-free so settings UI and tests share the same values.

/// Canonical hosted MCP endpoint path appended to the API base URL.
const String kMcpEndpointPath = '/v1/mcp';

/// OAuth Client ID registered for Claude's advanced connector settings.
const String kMcpOAuthClientId = 'omi-claude-prod';

/// Segment kinds for the syntax-highlighted Claude config block.
enum McpJsonToken { plain, key, string }

/// Canonical hosted MCP URL derived from [apiBaseUrl], with trailing slashes
/// normalized away so the endpoint is always joined exactly once.
String hostedMcpUrl(String apiBaseUrl) => '${apiBaseUrl.replaceAll(RegExp(r'/+$'), '')}$kMcpEndpointPath';

/// Ordered JSON segments of the Claude Code `~/.claude.json` `mcpServers`
/// entry. The copyable text and the highlighted code block are both built
/// from this single source so they can never drift apart.
List<(McpJsonToken, String)> hostedMcpConfigTokens(String mcpUrl) => [
      (McpJsonToken.plain, '{\n  '),
      (McpJsonToken.key, '"mcpServers"'),
      (McpJsonToken.plain, ': {\n    '),
      (McpJsonToken.key, '"omi"'),
      (McpJsonToken.plain, ': {\n      '),
      (McpJsonToken.key, '"type"'),
      (McpJsonToken.plain, ': '),
      (McpJsonToken.string, '"http"'),
      (McpJsonToken.plain, ',\n      '),
      (McpJsonToken.key, '"url"'),
      (McpJsonToken.plain, ': '),
      // jsonEncode keeps the output parseable even for an odd configured base.
      (McpJsonToken.string, jsonEncode(mcpUrl)),
      (McpJsonToken.plain, ',\n      '),
      (McpJsonToken.key, '"headers"'),
      (McpJsonToken.plain, ': {\n        '),
      (McpJsonToken.key, '"Authorization"'),
      (McpJsonToken.plain, ': '),
      (McpJsonToken.string, '"Bearer <key>"'),
      (McpJsonToken.plain, '\n      }\n    }\n  }\n}'),
    ];

/// The exact Claude Code `~/.claude.json` config JSON copied to the clipboard.
String hostedMcpConfigJson(String mcpUrl) => hostedMcpConfigTokens(mcpUrl).map((token) => token.$2).join();
